lib/route/link/bridge.c

```diff
@@ -1376,6 +1625,186 @@ struct rtnl_link_bridge_vlan *rtnl_link_bridge_get_port_vlan(struct rtnl_link *l
 	return NULL;
 }
 
+/**
+ * Set the Multiple Spanning Tree (MST) port state for a given MST instance
+ * @arg link		Link object of type bridge
+ * @arg instance	MST instance number
+ * @arg state		Port state to set (BR_STATE_*)
+ *
+ * @return 0 on success or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_NOMEM		Memory allocation failed
+ */
+int rtnl_link_bridge_set_mst_port_state(struct rtnl_link *link,
+					uint16_t instance, uint8_t state)
+{
+	struct bridge_data *bd;
+	struct mst_state_entry *existing_entry;
+	struct mst_state_entry *new_entry;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	existing_entry = find_mst_state_entry(bd, instance);
+
+	if (existing_entry != NULL) {
+		existing_entry->state = state;
+		return 0;
+	}
+
+	new_entry = mst_state_entry_create(instance, state);
+	if (new_entry == NULL)
+		return -NLE_NOMEM;
+
+	nl_list_add_tail(&new_entry->list_node, &bd->mst_list);
+	bd->ce_mask |= BRIDGE_ATTR_MST;
+
+	return 0;
+}
+
+/**
+ * Get the Multiple Spanning Tree (MST) port state for a given MST instance
+ * @arg link		Link object of type bridge
+ * @arg instance	MST instance number
+ *
+ * @return The state (BR_STATE_*) on success, or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_OBJ_NOTFOUND	MST instance not found
+ */
+int rtnl_link_bridge_get_mst_port_state(struct rtnl_link *link,
+					uint16_t instance)
+{
+	struct bridge_data *bd;
+	struct mst_state_entry *entry;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	entry = find_mst_state_entry(bd, instance);
+
+	if (entry == NULL)
+		return -NLE_OBJ_NOTFOUND;
+
+	return entry->state;
+}
+
+/**
+ * Delete the Multiple Spanning Tree (MST) port state for a given MST instance
+ * @arg link		Link object of type bridge
+ * @arg instance	MST instance number
+ *
+ * @return 0 on success or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_OBJ_NOTFOUND	MST instance not found
+ */
+int rtnl_link_bridge_del_mst_port_state(struct rtnl_link *link,
+					uint16_t instance)
+{
+	struct bridge_data *bd;
+	struct mst_state_entry *entry;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	entry = find_mst_state_entry(bd, instance);
+
+	if (entry == NULL)
+		return -NLE_OBJ_NOTFOUND;
+
+	mst_state_entry_del(entry);
+
+	if (nl_list_empty(&bd->mst_list))
+		bd->ce_mask &= ~BRIDGE_ATTR_MST;
```

Comment:
I don't think that the mask needs to be unset.

If you receive an object from kernel, then delete all entries and send it back, I think the fact that the `mst_list` is explicitly empty is important (and distinct from being unset).

Or not?

Same in `rtnl_link_bridge_clear_mst_port_state_info()`

Maybe calling delete should always explicitly set the flag on success (although here, we would expect that the flag was already set). This would be more a question for `rtnl_link_bridge_clear_mst_port_state_info()`.

Reply:
> If you receive an object from kernel, then delete all entries and send it back, I think the fact that the mst_list is explicitly empty is important (and distinct from being unset).

No, the kernel will reject an empty list being sent so we don't send the list if it's empty. Also, the list can just contain a subset of MST instances to modify the state of for the port; the ones not listed will just not be changed. Actually setting BRIDGE_ATTR_MST doesn't serve much purpose since we can just check if the list is non-empty, but I added it for consistency (maybe I could remove it?).

Reply:
ok, it's good then.

While BRIDGE_ATTR_MST is redundant (to just checking whether the list is empty), at various places we have also such flags when we could deduce via alternative ways whether the value is unset. I think it's fine to have it.